pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,576 @@
+// VolumeSnapshot, and then a VolumeSnapshotContent object is created to represent
+// this snapshot. After that, the controller will keep checking the snapshot status
+// though csi snapshot calls. When the snapshot is ready to use, the controller set
+// the status "Bound" to true to indicate the snapshot is bound and ready to use.
+// If the createtion failed for any reason, the Error status is set accordingly.
+// In alpha version, the controller not retry to create the snapshot after it failed.
+// In the future version, a retry policy will be added.
+
+const pvcKind = "PersistentVolumeClaim"
+const apiGroup = ""
+const snapshotKind = "VolumeSnapshot"
+const snapshotAPIGroup = crdv1.GroupName
+
+const controllerUpdateFailMsg = "snapshot controller failed to update"
+
+const IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+
+// syncContent deals with one key off the queue.  It returns false when it's time to quit.
+func (ctrl *csiSnapshotSideCarController) syncContent(content *crdv1.VolumeSnapshotContent) error {
+	klog.V(5).Infof("synchronizing VolumeSnapshotContent[%s]", content.Name)
+
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnShouldDelete) {
+		klog.V(5).Infof("syncContent[%s]: check ShouldDelete annotation on content.", content.Name)
+		switch content.Spec.DeletionPolicy {
+		case crdv1.VolumeSnapshotContentRetain:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Retain, nothing to do", content.Name)
+
+		case crdv1.VolumeSnapshotContentDelete:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Delete", content.Name)
+			ctrl.deleteSnapshotContent(content)
+			klog.V(5).Infof("syncContent: check if we should remove Finalizer for VolumeSnapshotContent[%s]", content.Name)
+			// It is a deletion candidate if DeletionTimestamp is not nil and
+			// VolumeSnapshotContentFinalizer is set.
+			if utils.IsContentDeletionCandidate(content) {
+				// Volume snapshot content is a deletion candidate.
+				// Remove the content finalizer.
+				klog.V(5).Infof("syncContent: Content [%s] is a deletion candidate. Check if it is bound to a snapshot.", content.Name)
+				return ctrl.removeContentFinalizer(content)
+			}
+
+		default:
+			// Unknown VolumeSnapshotDeletionolicy
+			ctrl.eventRecorder.Event(content, v1.EventTypeWarning, "SnapshotUnknownDeletionPolicy", "Volume Snapshot Content has unrecognized deletion policy")
+		}
+		// By default, we use Retain policy if it is not set by users
+		klog.V(4).Infof("VolumeSnapshotContent[%s]: the policy is %s", content.Name, content.Spec.DeletionPolicy)
+
+	} else {
+		var err error
+		if !ctrl.isDriverMatch(content) {
+			klog.V(3).Infof("syncContent: Should not provision snapshot for content %s. Driver does not match", content.Name)
+			return nil
+		}
+		klog.V(5).Infof("syncContent: Call CreateSnapshot for content %s", content.Name)
+		if content.Spec.Source.VolumeHandle != nil && content.Status == nil {
```

Comment:
what about content.Status.SnapshotHandle == nil?
there are cases where CreateSnapshot fails, then Status != nil, however the handle is not there yet
in this case, CreateSnapshot should be called to retry.

Reply:
In the previous design, we don't retry when create snapshot fails because snapshot is time sensitive.  Now we have another PR "https://github.com/kubernetes-csi/external-snapshotter/pull/153" which adds retries.  We can handle the retries in that PR after this one.